rb_ws/src/buggy/scripts/auton/rolling_sanity_check.py

```diff
@@ -0,0 +1,118 @@
+
+        self.filter_gps_status_publisher = rospy.Publisher(self_name + "/debug/filter_gps_seperation_status", Bool, queue_size=1)
+
+        self.covariance_status_publisher = rospy.Publisher(self_name + "/debug/covariance_status", Bool, queue_size=1)
+
+        self.error_message_publisher = rospy.Publisher(
+            self_name + "/nav/status/error_messages", String, queue_size=1
+        )
+
+        self.status_flags_publisher = rospy.Publisher(
+            self_name + "/nav/status/tripped_status_flags", Int8MultiArray, queue_size=1
+        )
+
+
+
+    def update_overrange_status(self, msg : ImuOverrangeStatus):
+        self.imu_overrange_status = msg.status
+
+    def update_status_flags(self, msg : FilterStatus):
+        self.status_flag_val = msg.gq7_status_flags
+
+    def update_gps_location(self, msg : PoseStamped):
+        self.gps_location = msg.pose
+
+    def update_filter_location(self, msg):
+        self.filter_location = msg.pose
+
+    def calc_covariance(self):
+        if (self.filter_location == None):
+            self.covariance_status_publisher.publish(False)
+        else:
+            self.covariance_status_publisher.publish(self.filter_location.covariance[0] ** 2 + self.filter_location.covariance[7] ** 2 <= 1**2)
+
+    def calc_locations(self):
+        # currently comparing cross-track error, checking less than 0.5 m
+        if (self.filter_location == None or self.gps_location == None):
+            self.filter_gps_status_publisher.publish(False)
+        else:
+            self.filter_gps_status_publisher.publish(abs(self.filter_location.pose.position.y - self.gps_location.position.y) < 0.5)
+
+    def is_overrange (self):
+        s = self.imu_overrange_status
+        if (s == None):
+            self.overrange_status_publisher.publish(False)
+        else:
+            accel_status = s.status_accel_x or s.status_accel_y or s.status_accel_z
+            gyro_status = s.status_gyro_x or s.status_gyro_y or s.status_gyro_z
+            mag_status = s.status_mag_x or s.status_mag_y or s.status_mag_z
+
+            # publishes true if ALL flags are FALSE (if the imu looks good)
+            self.imu_overrange_status.publish(not (accel_status or gyro_status or mag_status or s.status_press))
+
+    def filter_status_warning (self):
+        b = bin(self.status_flag_val)
+        self.flags = Int8MultiArray()
+        self.flags.data = []
+        error_message = ""
+        for i in range (len(b)):
+            if (b[i] == '1'):
+                self.flags.append(i)
+                error_message += self.error_messages[i] + " "
+
+        # publish bit values
+        self.status_flags_publisher.publish(self.flags)
+
+        # publish string translations
+        self.error_message_publisher.publish(error_message)
+
+    def sanity_check(self):
+        self.calc_covariance()
+        self.calc_locations()
+        self.is_overrange()
+        self.filter_status_warning()
+
```

Comment:
can you add an overall check, where if any of these checks fail, a warning is published

Reply:
(i added this but) would it be helpful considering some of the flags get tripped occasionally just as the buggy is rolling normally? i don't see how it would help us determine if/when something unusual is happening.

Reply:
I think the point is, if any of these flags get tripped for longer than, say, a second, the human driver should intervene. We have one warning light on the buggy, so there's really only one topic that firmware can take. IMO filter divergence should warrant immediate intervention, but sustained covariance, overrange or fitler status warnings should also warrant intervention. What are your thoughts on the most appropriate criteria for intervention?

Reply:
Maybe publish the alarm light topic as an int, 0 if all is good, 1 if anything is not normal, and 2 if filter diverges? Then I can ask firmware to send corresponding colors

Reply:
Do we want to alert the human driver if the filter separates briefly but reconverges? Also, I think checking if any flags are tripped for some sustained period of time makes sense - you've looked at the bags more than I have, do you think one second is significant for all of them?